### DISCUSSION TITLE
Auth for bulk delete in table

### DISCUSSION THREAD

**Discussion by vv-ac** [filamentphp/filament#2338] [Help] [answered]
Hi all,

I have a resource with a table that has mulitple entries.
Deleting them should only be possible for a few, so I set up a policy for the model:
```
    public function delete(User $user, Resource $resource
    {
        if($resource->is_default != true) {
            $resources = $user->resources;

            foreach($resources as $resource{
               ...
            }
        }
    }
```

Inside the resource class I added the following:
```
    protected function getTableBulkActions(): array
    {
        return [
            BulkAction::make('delete')
                ->action(fn (Collection $records) => $records->each->delete())
                ->deselectRecordsAfterCompletion()
                ->hidden(fn (Resource $record): bool => auth()->user()->can('delete', $record))
        ];
    }
```
Here, it does not matter if I go for "can", "cannot", "Resource $record" or "Collection $records", I still can delete the entries in the table via marking the checkbox in front of the rows and then delete them.

Any idea what is wrong with my approach?

Thanks!

**Comment by danharrin** [comment, marked as answer]
Your policy method only validates a single record, but you're trying to pass a collection of them into it from the `hidden()` method. Bulk actions were designed to run entirely on the frontend, so it's not yet possible to dictate whether the bulk action is hidden only when certain records are selected. If this was possible, I would be concerned about the possible performance impact of looping through every record checking a policy. This is why we use the `deleteAny` policy method for the built-in admin bulk delete action, instead of the standard `delete`.

**Reply by vv-ac** [reply]
Thank you.
I simply return a false now for deleteAny in my policy which disables the bulkAction in total.

Also, I understand that it is likely to result in an performance issue when inspecting all elements for certain attributes, maybe I'll later think of some deferred method to check for such.
